### ISSUE TITLE
Update CI for deployment policies under semantic versioning

### ISSUE THREAD

**Issue by rileyjmurray** [cvxpy/cvxpy#1569]
Today marked the first-ever cvxpy developer call! During the call we announced an intention to adopt semantic versioning (semver.org) and a regular release schedule for cvxpy.

The deployment phase of our CI will need to be updated to reflect our plans w.r.t. semantic versioning. The easiest change is that GitHub Actions deployments need to be turned off on the master branch.

A new branch will be created for each minor release. Development should never occur directly on one of these branches. Rather, if a minor release needs a patch, then you can branch and create a pull request to merge back into the release branch. I suggest these release branches are named things like "maintenance/1.1.x" and "maintenance/1.2.x" (where the character x is used literally). That's the convention used by NumPy and SciPy. @SteveDiamond @akshayka does that sound alright to you two?

For now, I'll assume that we use that naming scheme. Our deployments should run only on branches "maintenance/1.v.x" (for some integer v) when the patch version number has been incremented in ``cvxpy.__version__``. Or, if we're careful about merging commits onto those branches, we can presume that the version number is updated at every commit and hence deploy on every commit. @phschiele  what are your thoughts on how to do this? If we only support a handful of minor releases at any given time then my guess is that we can list them explicitly in the configuration for GitHub Actions.

**Comment by phschiele** [comment]
@rileyjmurray @akshayka From the CI perspective, I think not a lot has to be changed.
To always run the tests on any push to these new "branch types", we only need to extend the `on` section to include pushes to `maintenance/*`/`release/*`

The current logic for when to deploy is only based on tags (see below), so if we go ahead with Akshay's suggestion, the currently supported versions do not need to be maintained on the CI level (but can of course be documented/communicated elsewhere).

Deploy when: `[[ $GITHUB_EVENT_NAME == 'push' && $GITHUB_REF == 'refs/tags'* ]]`

**Comment by rileyjmurray** [comment]
I forgot that deployments were already based on commit tags. Glad to hear that moving forward will be simple, @phschiele!

@akshayka I suggested ``maintenance/1.v.x`` just because that's the convention used by numpy and scipy. But I went back and looked at other NumFOCUS projects, and it seems that convention is far from universal. I'm on board with using ``release/1.v.x``.

Let's start to make plans for the 1.2 release. I suggest we create a branch called ``release/1.1.x`` that starts from [this commit](https://github.com/cvxpy/cvxpy/commit/b275680485117eb28c3e2f193920119443ad9d6a). Does that sound alright to you @akshayka @SteveDiamond @bstellato? From there we'd be treating "master" as a pre-release of 1.2. (I'd like to figure out a policy for how we maintain cvxpy version numbers on the master branch. For example, if someone installs from master, they shouldn't see v1.1.17 since that's already been released. I'll look into this and come back with a proposal / PR.)
